N°3677 - Fix AttributeImage.default_image URLs not up to date after app_root_url change

piRGoif opened the pull request:
## Original work
This is a cherry pick of commits made in #176 (original author @sg-gcouronne , many thanks to him !)
The original PR was looked at far too late, so we had no choice but merging manually the changes to a new branch, based on a more recent code.

The original PR aimed at fixing the mandatory run of a setup after changing the `app_root_url` config parameter. There were 3 issues addressed, changing absolute URLs generated during setup on : 
* class icon
* display template
* AttributeImage default_image


## Two concerns already fixed
Since iTop 3.0.0 : 
* the icon compilation is now done in MFCompiler::GenerateStyleDataFromNode, which creates a call to ormStyle passing the relative path to the icon (in other words : absolute url computation is done at runtime)
* display template were removed (N°3233)


## Remaining problem on AttributeImage default_image
Examples in the default datamodel : Person.picture (images/silhouette.png) and admin Person instance.

### To reproduce (on Windows) : 

1. create a new lmhosts entry
2. run setup using an URL with this entry
3. remove the entry (comment lmhosts declaration + run `nbtstat -R`)
4. change app_root_url config param to localhost (DO NOT run setup again !)
5. access the previously installed iTop instance with localhost

Before change : 
Broken default image in the admin Person instance

After change : 
Default image in admin Person instance is OK


## TODO

- [x] Change Test annotations
- [x] Rebase on support/3.2
- ~~Change since annotations~~
- [x] Document AttrimageImage.default_image path is relative to the env root (note that an AttributeImage was added in Action core class : N°2039)

piRGoif committed:
Apply #176 modifications

piRGoif committed:
Fix Undefined variable $aParams

piRGoif committed:
Revert modification for class icon, as it is now computed at runtime via ormStyle

Molkobain commented:
Technical review:
- Remove AttributeBlob::__construct() as we don't know why it's there.
- Then it should be ok

Molkobain commented on core/attributedef.class.inc.php:
```diff
@@ -8626,9 +8626,36 @@ class AttributeImage extends AttributeBlob
 	}
 
+	/**
+	 * {@inheritDoc}
```
```suggestion
	 * @inheritDoc
```

piRGoif replied:
Actually the phpdoc block doesn't give useful information at all. I'll remove it.

piRGoif replied:
Done in ee2f3765

piRGoif committed:
:bulb: Remove useless \AttributeImage::Get phpdo

piRGoif committed:
Remove \AttributeImage::__construct absolute url init
Already made in the setup, and if a user get the new code it is because of a version update so the setup was run

piRGoif commented:
> * Remove AttributeBlob::__construct() as we don't know why it's there.

Done in 6ba7c450 (it was AttributeImage constructor by the way)

Molkobain reviewed (changes requested):
Can you add a unit test to ensure that `\AttributeImage::Get()` return the correct URL for the default image even when we change the `app_root_url`?

There is the `Person->picture` attribute in the standard DM that you can use.

piRGoif committed:
Small changes in \AttributeImage::Get

piRGoif committed:
New test for AttributeImage::Get on default_image

piRGoif commented:
> Can you add a unit test to ensure that \AttributeImage::Get() return the correct URL

Very good idea ! 
Done in bb621f9f

Note that there was a cache in \utils::GetAbsoluteUrlAppRoot (called from \AttributeImage::Get) that cannot be cleared from the test :( I removed it completely, I don't think there should be any perceptible drawback as everything is done in memory ? To be discussed during technical review...

piRGoif committed:
Restore cache in \utils::GetAbsoluteUrlAppRoot
Was needed by utilsTest::testGetAbsoluteUrlAppRootPersistency

piRGoif committed:
Woops keep the cache only where it belongs
Now all tests are ok O:)

piRGoif commented:
Well, as removing the cache made utilsTest::testGetAbsoluteUrlAppRootPersistency to crash, I restored it but partially : only for the use case of the test (having app_root_url empty)

Molkobain commented:
Accepted during functional review in 3.2 because not much requestors, low impact and workaround (run setup again).
We just have to check the utils method, then merge.

piRGoif committed:
Fix AttributeImageTest::testGet VS \utils::GetAbsoluteUrlAppRoot cache without changing GetAbsoluteUrlAppRoot algo

piRGoif commented on tests/php-unit-tests/unitary-tests/core/AttributeImageTest.php:
```diff
@@ -0,0 +1,48 @@
+ * @runTestsInSeparateProcesses
+ * @preserveGlobalState disabled
+ * @backupGlobals disabled
+ */
```
To remove before merge !

piRGoif replied:
Done !

Molkobain commented:
Support PRs review:
- Ok with the refactoring
- Since iTop 3.2 and N°2039, we have an AttributeImage in the datamodel.core.xml, ok to keep path relative to env-xxx. Pierre will just update the documentation to make it clear that these paths must be relative to tne current environment and not the app root.

piRGoif committed:
Remove merge error

piRGoif committed:
AttributeImageTest : remove test class annotations
No need for processIsolastion anymore

piRGoif committed:
Woops wasn't a merge error but an undocumented test only attribute

This reverts commit 6ef98443378eb11aaa2cbfa71138a4629030fa7b.

piRGoif committed:
Fix utilsTest::testGetAbsoluteUrlAppRootPersistency as we are now using a static class attribute

piRGoif committed:
Remove now useless var
